### PR TITLE
SSL Support For the API

### PR DESCRIPTION
**Pull request by breakthesec**
Added SSL Support to the API.  Automatically creates Self-Signed Certificate and places the files under ~/.w3af/ssl/ , if the cert file not present. User can disable ssl support with "--no-ssl"

### CODE REVIEW

File: w3af/core/ui/api/main.py [hunk 1]
```diff
@@ -24,17 +24,13 @@
```
Comment (andresriancho) [inline w3af/core/ui/api/main.py]: We don't need the exec permissions for this file `100644 → 100755`, please move back to 644

File: w3af/core/ui/api/main.py [hunk 2]
```diff
@@ -24,17 +24,13 @@
...
+from w3af.core.ui.api.utils.digital_certificate import SSLCertifiacate
 
 def main():
     """
     Entry point for the REST API
     :return: Zero if everything went well
     """
-    # Check if I have all needed dependencies
-    dependency_check()
-
```
Comment (andresriancho) [inline w3af/core/ui/api/main.py]: Why this change? I believe you did something wrong with the merge/pull request. You should base your REST API changes on my develop, and then send a PR to the develop branch.

File: w3af/core/ui/api/utils/digital_certificate.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
+from w3af.core.controllers.misc.homeDir import get_home_dir
+import os
+
+from OpenSSL import crypto
+import socket,ssl
+
+class SSLCertifiacate:
+    def __init__(self):
+        ssl_dir = os.path.join(get_home_dir(),"ssl")
```
Comment (andresriancho) [inline w3af/core/ui/api/utils/digital_certificate.py]: Please respect PEP-8 , after a ',' there should be a space.

Same applies to all the other pep-8 related stuff in your change.

File: w3af/core/ui/api/utils/digital_certificate.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+            cert.get_subject().L = "w3af"
+            cert.get_subject().O = "W3AF"
+            cert.get_subject().OU = "W3AF"
+            cert.get_subject().CN = host
+            cert.set_serial_number(1111)
+            cert.gmtime_adj_notBefore(0)
+            cert.gmtime_adj_notAfter(10*365*24*60*60)
+            cert.set_issuer(cert.get_subject())
+            cert.set_pubkey(key)
+            cert.sign(key, "sha1")
```
Comment (andresriancho) [inline w3af/core/ui/api/utils/digital_certificate.py]: Some browsers/clients are deprecating SSL certs signed with sha1, please use a stronger proto

**Comment by breakthesec** [comment]
I've updated the code.  Changed SSL-signining algorithm from sha1 to sha256 and made other changes.

File: w3af/core/ui/api/utils/digital_certificate.py [hunk 5]
```diff
@@ -0,0 +1,46 @@
+from w3af.core.controllers.misc.homeDir import get_home_dir
+import os
+
+from OpenSSL import crypto
+import socket, ssl
+
+
+class SSLCertifiacate:
```
Comment (andresriancho) [inline w3af/core/ui/api/utils/digital_certificate.py]: Typo `SSLCertifiacate`

Also, should inherit from `object`

File: w3af/core/ui/api/utils/cli.py [hunk 6]
```diff
@@ -61,6 +61,12 @@ def parse_arguments():
                         default=False,
                         nargs='?')
 
+    parser.add_argument('--no-ssl',
+                        dest="disableSSL",
```
Comment (andresriancho) [inline w3af/core/ui/api/utils/cli.py]: Use `disable_ssl`, all other options are underscore separated

File: w3af/core/ui/api/main.py [hunk 7]
```diff
@@ -43,8 +45,13 @@ def main():
...
     try:
-        app.run(host=app.config['HOST'], port=app.config['PORT'],
-                debug=args.verbose, use_reloader=False, threaded=True)
+
+        if args.disableSSL:
+            app.run(host=app.config['HOST'], port=app.config['PORT'],
+                    debug=args.verbose, use_reloader=False, threaded=True)
+        else:
+            app.run(host=app.config['HOST'], port=app.config['PORT'],
+                    debug=args.verbose, use_reloader=False, threaded=True, ssl_context=SSLCertifiacate().get())
```
Comment (andresriancho) [inline w3af/core/ui/api/main.py]: Improvement: Call `get(app.config['HOST'])` and change `get` to forward the host to `generate` so the user provided host is used in the SSL cert creation.

File: w3af/core/ui/api/utils/digital_certificate.py [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+                f.write(crypto.dump_certificate(crypto.FILETYPE_PEM, cert))
+
+            with open(self.key_path, "w") as f:
+                f.write(crypto.dump_privatekey(crypto.FILETYPE_PEM, key))
+
+    def get(self):
+        if not os.path.exists(self.cert_path) or not os.path.exists(self.cert_path):
+            self.generate()
+        context = ssl.SSLContext(ssl.PROTOCOL_TLSv1_2)
+        context.load_cert_chain(self.cert_path, self.key_path)
```
Comment (andresriancho) [inline w3af/core/ui/api/utils/digital_certificate.py]: Users will want to use their own certificates, please add the configuration parameters to allow them to do so.

File: w3af/core/ui/api/utils/cli.py (hunk 6 above)
Comment (andresriancho) [inline w3af/core/ui/api/utils/cli.py]: Please check [these lines](https://github.com/andresriancho/w3af/blob/master/w3af/core/ui/api/utils/cli.py#L185-L187) , I believe the disable ssl setting should also show/hide that message.

**Comment by andresriancho** [comment]
@breakthesec ping! Are you going to address the comments/issues I made a while ago?

**Comment by andresriancho** [comment]
Merged your changes into develop with some improvements and documentation.

@breakthesec thanks for the initial code!
